File: baekjoon_all/01000+/boj_1967.cpp
// Solve 2023-02-13
// Update 2025-10-04

#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);
#define ALL(v) v.begin(),v.end()
#define UNIQUE(v) v.erase(unique(v.begin(),v.end()),v.end());
#define SETW(n, c) cout << setw(n) << setfill(c);
#define SETP(n) cout << fixed << setprecision(n);

using namespace std;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;
using ld = long double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;
const int INF = 0x3f3f3f3f;
const int MOD = 1000000007;

pii dfs(int cur, vector<vector<pii>>& tree) {
    vector<int> max_depth_list = {0, 0};
    int max_len = 0;

    for (pii& p : tree[cur]) {
        int nxt = p.first;
        int w = p.second;

        pii nxt_res = dfs(nxt, tree);
        max_depth_list.push_back(nxt_res.first + w);
        max_len = max(max_len, nxt_res.second);
    }

    sort(max_depth_list.begin(), max_depth_list.end(), greater<int>());
    return {max_depth_list[0], max(max_depth_list[0] + max_depth_list[1], max_len)};
}

int main() {
    FASTIO;

    int n;
    cin >> n;
    vector<vector<pii>> tree(n + 1);

    for (int i = 1; i < n; i++) {
        int par, child, w;
        cin >> par >> child >> w;
        tree[par].emplace_back(child, w);
    }

    pii res = dfs(1, tree);
    cout << res.second << '\n';
    return 0;
}

File: baekjoon_all/01000+/boj_1967.py
# Solve 2025-10-04

import sys
sys.setrecursionlimit(10 ** 6)

input = lambda: sys.stdin.readline().rstrip()

def dfs(cur):
    max_depth_list = [0, 0]
    max_len = 0

    for nxt, w in tree[cur]:
        d, l = dfs(nxt)
        max_depth_list.append(d + w)
        max_len = max(max_len, l)

    max_depth_list.sort(reverse=True)
    return max_depth_list[0], max(max_depth_list[0] + max_depth_list[1], max_len)


n = int(input())
tree = [[] for _ in range(n + 1)]

for _ in range(1, n):
    par, child, w = map(int, input().split())
    tree[par].append((child, w))

max_depth, max_len = dfs(1)
print(max_len)
